src/lib/uri/uri.c

```diff
@@ -9,9 +9,115 @@
 
 #define XSTRNDUP(s, n)  (s != NULL ? xstrndup(s, n) : NULL)
 
+/**
+ * Find URI query param by its @a name in @a uri structure.
+ */
+static struct uri_query_param *
+uri_find_query_param(const struct uri *uri, const char *name)
+{
+	for (int i = 0; i < uri->param_count; i++) {
+		if (strcmp(uri->params[i].name, name) == 0)
+			return &uri->params[i];
+	}
+	return NULL;
+}
+
+/**
+ * Add new @a value to URI query @a param.
+ */
+static void
+uri_query_param_add_value(struct uri_query_param *param, const char *value)
+{
+	size_t size = (param->value_count + 1) * sizeof(char *);
+	param->values = xrealloc(param->values, size);
+	param->values[param->value_count++] = xstrdup(value);
+}
+
+/**
+ * Destroy URI query @a param and free all associated resourses.
+ */
+static void
+uri_query_param_destroy(struct uri_query_param *param)
+{
+	for (int i = 0; i < param->value_count; i++)
+		free(param->values[i]);
+	free(param->values);
+	free(param->name);
+	TRASH(param);
+}
+
+/**
+ * Create new URI query @a param, with given @a name.
+ */
+static void
+uri_query_param_create(struct uri_query_param *param, const char *name)
+{
+	memset(param, 0, sizeof(struct uri_query_param));
```

Comment:
Nit: there are just two parameters you need to zero. I'd set them explicitly without using memset - it'd be more readable IMO:

```C
param->name = xstrdup(name);
param->values = NULL;
param->value_count = 0;
```

Same comment for `uri_set_create`. Up to you.

Reply:
fixed